### PR TITLE
Add support generic cape function signature (more than one input and output)

### PR DESCRIPTION
**Pull request by yanndupis**
This PR add the following:
- Support functions with multiple inputs and outputs (need to use `@serdio.lift_io` to work).
- multiple inputs can be specified in cape.run and cape.invoke as args or kwargs as follow: `cape.run(function_id, x, y, use_serdio=True)` or `cape.run(function_id, x=x, y=y, use_serdio=True)`.
- Fix arg in `serdio.deserialize` and `serdio.serialize` (in cape.py).

I tested against the enclave with single argument and and multiple arguments (with` @serdio.lift_io`) and it worked properly.

[CAPE-729](https://capeprivacy.atlassian.net/browse/CAPE-729)

### CODE REVIEW

File: pycape/cape.py [hunk 1]
```diff
@@ -90,23 +93,33 @@ async def _connect(self, function_id):
 
         return
 
-    async def _invoke(self, input, serde_hooks, use_serdio):
+    async def _invoke(self, *args, serde_hooks, use_serdio, **kwargs):
```
Comment (jvmncs) [inline pycape/cape.py]: nit: I think putting `serde_hooks` & `use_serdio` before `*args` in each of these private methods would be more conventional

File: serdio/serdio/io_lifter.py [hunk 2]
```diff
@@ -160,7 +161,14 @@ def cape_handler(input_bytes):
                     "Make sure your input is serialized with MessagePack manually or "
                     "by setting use_serdio=True in Cape.run or Cape.invoke"
                 )
-            output = self._func(f_input)
+
+            if _check_if_multiple_inputs(f_input):
+                args, kwargs = f_input.get("args"), f_input.get("kwargs")
```
Comment (jvmncs) [inline serdio/serdio/io_lifter.py]: nit: seems like here we do `f_input.get("args")` and `f_input.get("kwargs")` twice -- once in `_check_if_multiple_inputs`, then once again immediately afterwards. seems like it would be better to refactor `_check_if_multiple_inputs` to return the args/kwargs. then your conditional could be
```python
args, kwargs = _check_inputs(f_input)
if args is None and kwargs is None:
    output = self._func(f_input)
else:
    ...
```
and you will have only checked for those key-value pairs in `f_input` one time

File: pycape/cape.py [hunk 3]
```diff
@@ -90,23 +93,33 @@ async def _connect(self, function_id):
...
+        # If multiple args and kwargs are supplied to the function, bundle
+        # them into a dictionary before serializing, so can be deserialized
+        # and supplied to cape handler properly.
+        if len(args) == 1 and len(kwargs) == 0:
+            inputs = args[0]
+        elif len(args) == 0 and len(kwargs) == 1:
+            inputs = kwargs.popitem()[1]
+        else:
+            inputs = {"args": args, "kwargs": kwargs}
```
Comment (jvmncs) [inline pycape/cape.py]: think I see an issue with this, in particular reserving this type of input for ourselves could lead to danger for the unaware user.

let's say I write a function that requires an analogous dictionary, e.g.
```python
@serdio.lift_io
def my_func(d):
    args = d["args"]
    kwargs = d["kwargs"]
    return do_something(args, kwargs)
    
cape_handler = my_func.as_cape_handler()
```
in my run script I use this function like so:
```python
my_dict = {"args": (1.0, 3.0), "kwargs": {"do_something_kwarg": 5.0}}
cape.invoke(my_dict, use_serdio=True)
```

due to this construction and the logic in [IOLifter](https://github.com/capeprivacy/pycape/pull/37/files#diff-1554a382ecdf23501c04e4dc84e8b5358a3da65a885121a8de3ce86e49128b93R164-R171) below, I think the cape handler used by sentinel will try to unwrap this inner dictionary `d` and pass that to `my_func`. so for example you'de end up with sentiel calling the following via its launch.py script:
```python
my_func(1.0, 3.0, do_something_kwarg=5.0)
```
which would of course fail because `my_func` expects a dict that it then unwraps and passes to `do_something`

Comment (jvmncs) [reply pycape/cape.py]: I think there are two issues at play:
1. we need to establish a contract between pycape & serdio dictating how pycape will pack up args for the `IOLifter`'s cape_handler to unpack. right now, pycape is declaring the contract in these lines of code, and serdio is following it in the `IOLifter.as_bytes_handler` method. however, serdio is a dependency of pycape and therefore precedes it, so really serdio should be the one declaring the contract, and pycape should be the one following it.
2. currently, this contract is a dictionary `{"args": <tuple_args>, "kwargs": <dict_kwargs>}`. this form of the contract is a bit too easy for a user to accidentally reproduce in their own use case. 

- to fix (2), I think we can just make the keys of this dictionary a bit more specific, so that it's unlikely a user will stumble into using the exact same dictionary keys for their own applications
- to fix (1), I think the exact structure of this dictionary should be determined by serdio. this means moving all functionality related to args/kwargs packing & unpacking into serdio and exposing it as a utility function, which pycape can then use in `Cape._invoke`

**Review by jvmncs — COMMENTED** [review]
think there is one issue that should be fixed, other comments are just nits

File: pycape/cape.py (hunk 3 above)
Comment (yanndupis) [reply pycape/cape.py]: This is good feedback! I am going to look into it. thanks

**Comment by yanndupis** [comment]
Hey @jvmncs, it should be ready for another review. thanks!

File: serdio/serdio/utils.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
+def pack_function_args_kwargs(args, kwargs):
```
Comment (jvmncs) [inline serdio/serdio/utils.py]: nit: `utils.py` is asking for unrelated code to make its way in here. maybe rename the file to `func_utils.py` or something else?

**Review by jvmncs — APPROVED** [review]
LGTM, one nit

File: serdio/serdio/utils.py (hunk 4 above)
Comment (yanndupis) [reply serdio/serdio/utils.py]: good point, thanks

**Comment by yanndupis** [comment]
Thanks for the review again!
